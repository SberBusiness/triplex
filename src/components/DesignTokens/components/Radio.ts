import {TDesignTokenValue, TDesignTokenValues} from '@sberbusiness/triplex/components/DesignTokens/types/DesignTokenTypes';

// Название токенов компонента Radio.
export const designTokensComponentsRadioKeys = [
    'Background_Checked_Default',
    'Background_Checked_Hover',
    'Background_Default',
    'Background_Disabled',
    'BorderColor_Default',
    'BorderColor_Checked',
    'BorderColor_Disabled',
    'BorderColor_Focused',
    'BorderColor_Hover',
    'Color_Default',
    'Color_Disabled',
    'Dot_Default',
    'Dot_Disabled',
] as const;
// Тип, содержащий названия токенов компонента Radio.
export type TDesignTokensComponentsRadioKeys = (typeof designTokensComponentsRadioKeys)[number];
// Тип, содержащий названия токенов компонента Radio и их значения.
export type TDesignTokensComponentsRadioValue = Record<TDesignTokensComponentsRadioKeys, TDesignTokenValue>;
// Тип, содержащий названия токенов компонента Radio и их значения в светлой и темной теме.
export type TDesignTokensComponentsRadioValues = Record<TDesignTokensComponentsRadioKeys, TDesignTokenValues>;
// Тип локальных токенов компонента Radio.
export type TDesignTokensComponentsRadio = {Radio: TDesignTokensComponentsRadioValue};

// Токены компонента Radio в светлой и темной темах.
export const Radio_Tokens: TDesignTokensComponentsRadioValues = {
    Background_Checked_Default: [{ref: 'Primary.700'}, {ref: 'Primary.500'}], // var(--triplex-Radio-Background_Checked_Default)
    Background_Checked_Hover: [{ref: 'Primary.500'}, {ref: 'Primary.700'}], // var(--triplex-Radio-Background_Checked_Hover)
    Background_Default: [{ref: 'Basic.100'}, {ref: 'Basic.0'}], // var(--triplex-Radio-Background_Default)
    Background_Disabled: [{ref: 'Neutral.100'}, {ref: 'Basic.500'}], // var(--triplex-Radio-Background_Disabled)
    BorderColor_Checked: [{ref: 'Primary.700'}, {ref: 'Primary.500'}], // var(--triplex-Radio-BorderColor_Checked)
    BorderColor_Default: [{ref: 'Neutral.500'}, {ref: 'Basic.300'}], // var(--triplex-Radio-BorderColor_Default)
    BorderColor_Disabled: [{ref: 'Neutral.500'}, {ref: 'Basic.300'}], // var(--triplex-Radio-BorderColor_Disabled)
    BorderColor_Focused: [{ref: 'Accent.500'}, {ref: 'Accent.500'}], // var(--triplex-Radio-BorderColor_Focused)
    BorderColor_Hover: [{ref: 'Primary.500'}, {ref: 'Primary.700'}], // var(--triplex-Radio-BorderColor_Hover)
    Color_Default: [{ref: 'Basic.700'}, {ref: 'Neutral.100'}], // var(--triplex-Radio-Color_Default)
    Color_Disabled: [{ref: 'Basic.700'}, {ref: 'Basic.300'}], // var(--triplex-Radio-Color_Disabled)
    Dot_Default: [{ref: 'Basic.100'}, {ref: 'Neutral.300'}], // var(--triplex-Radio-Dot_Default)
    Dot_Disabled: [{ref: 'Neutral.500'}, {ref: 'Basic.300'}], // var(--triplex-Radio-Dot_Disabled)
};
