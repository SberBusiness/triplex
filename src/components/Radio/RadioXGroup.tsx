import React from 'react';
import {TIndentSize} from '@sberbusiness/triplex/consts/IndentConst';
import {classnames} from '@sberbusiness/triplex/utils/classnames/classnames';
import {cssClass} from '@sberbusiness/triplex/utils/cssClass';

/** Свойства компонента RadioXGroup. */
export interface IRadioXGroupProps extends React.HTMLAttributes<HTMLDivElement> {
    /** Размер отступа. */
    indent?: TIndentSize;
}

/** Группа радио-кнопок с направлением по оси X. */
export const RadioXGroup: React.FC<IRadioXGroupProps> = (props) => {
    const {children, className, indent = 12, ...rest} = props;
    const classNames = classnames('cssClass[radioXGroup]', cssClass(`indent-${indent}`), className);

    return (
        <div className={classNames} role="radiogroup" {...rest}>
            {children}
        </div>
    );
};

RadioXGroup.displayName = 'RadioXGroup';
