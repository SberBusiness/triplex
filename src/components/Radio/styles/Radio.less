@import 'src/styles/style';

.label {
    display: inline-block;
    position: relative;
    min-width: @ufs-theme-size-icon-16;
    min-height: @ufs-theme-size-icon-16;
    font-weight: 400;
    font-size: 14px;
    line-height: 20px;
    color: var(--triplex-Radio-Color_Default);
    cursor: pointer;

    &.nonempty {
        padding-left: 24px;
    }

    &.disabled {
        color: var(--triplex-Radio-Color_Disabled);
        cursor: default;
    }
}

.radio {
    margin: 0;
    opacity: 0;
}

.radioIcon {
    box-sizing: border-box;
    background: var(--triplex-Radio-Background_Default);
    border: 1px solid var(--triplex-Radio-BorderColor_Default);
    border-radius: 50%;
}

.radioIcon::after {
    content: '';
    position: absolute;
    width: 6px;
    height: 6px;
    top: 4px;
    left: 4px;
    background: var(--triplex-Radio-Dot_Default);
    border-radius: 50%;
    transform: scale(0);
}

.radioIcon,
.radioIcon::after {
    pointer-events: none;
}

.radio,
.radioIcon {
    position: absolute;
    width: @ufs-theme-size-icon-16;
    height: @ufs-theme-size-icon-16;
    left: 0;

    .nonempty & {
        top: 2px;
    }
}

.radio:enabled {
    cursor: pointer;
}

.radio:checked {
    + .radioIcon {
        background: var(--triplex-Radio-Background_Checked_Default);
        border-color: var(--triplex-Radio-BorderColor_Checked);
    }

    + .radioIcon::after {
        transform: scale(1);
        transition: transform 200ms ease-out;
    }
}

.radio:hover + .radioIcon {
    border-color: var(--triplex-Radio-BorderColor_Hover);
}

.radio:disabled {
    + .radioIcon {
        background: var(--triplex-Radio-Background_Disabled);
        border-color: var(--triplex-Radio-BorderColor_Disabled);
    }

    + .radioIcon::after {
        background: var(--triplex-Radio-Dot_Disabled);
    }
}

.radio[data-focus-visible]:focus + .radioIcon {
    border-color: var(--triplex-Radio-BorderColor_Focused);
}

.radio:checked:enabled:hover + .radioIcon {
    background-color: var(--triplex-Radio-Background_Checked_Hover);
}
