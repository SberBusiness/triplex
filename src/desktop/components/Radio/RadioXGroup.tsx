import React from 'react';
import {TIndentSize} from '@sberbusiness/triplex/desktop/common/consts/IndentConst';
import {classnames} from '@sberbusiness/triplex/common/utils/classnames/classnames';
import {cssClass} from '@sberbusiness/triplex/desktop/utils/cssClass';

/** Свойства компонента RadioXGroup. */
export interface IRadioXGroupProps extends React.HTMLAttributes<HTMLDivElement> {
    /** Размер отступа. */
    indent?: TIndentSize;
}

/** Группа радио-кнопок с направлением по оси X. */
export const RadioXGroup: React.FC<IRadioXGroupProps> = (props) => {
    const {children, className, indent = 12, ...rest} = props;
    const classNames = classnames('cssClass[radioXGroup]', cssClass(`indent-${indent}`), className);

    return (
        <div className={classNames} role="radiogroup" {...rest}>
            {children}
        </div>
    );
};

RadioXGroup.displayName = 'RadioXGroup';
