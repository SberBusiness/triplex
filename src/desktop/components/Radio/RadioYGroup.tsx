import React from 'react';
import {classnames} from '@sberbusiness/triplex/common/utils/classnames/classnames';

/** Свойства компонента RadioYGroup. */
export interface IRadioYGroupProps extends React.HTMLAttributes<HTMLDivElement> {}

/** Группа радио-кнопок с направлением по оси Y. */
export const RadioYGroup: React.FC<IRadioYGroupProps> = (props) => {
    const {children, className, ...rest} = props;
    const classNames = classnames('cssClass[radioYGroup]', className);

    return (
        <div className={classNames} role="radiogroup" {...rest}>
            {children}
        </div>
    );
};

RadioYGroup.displayName = 'RadioYGroup';
