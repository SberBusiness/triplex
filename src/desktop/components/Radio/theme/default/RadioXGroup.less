@import 'src/desktop/common/theme/default';

@indent: 12, 16, 20, 24, 28, 32;

.radioXGroup {
    display: inline-flex;
    flex-flow: row wrap;

    each(@indent, {
        &.indent-@{value} {
            margin-left: unit(-@value / 2, px);
            margin-right: unit(-@value / 2, px);

            .label {
                margin-left: unit(@value / 2, px);
                margin-right: unit(@value / 2, px);
            }
        }
    });
}
