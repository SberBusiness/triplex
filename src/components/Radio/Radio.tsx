import React, {useState, useRef} from 'react';
import {IRadioProps} from '@sberbusiness/triplex/components/Radio/types';
import {EFocusSource} from '@sberbusiness/triplex/enums/EFocusSource';
import {classnames} from '@sberbusiness/triplex/utils/classnames/classnames';

/** Радио-кнопка с описанием. */
export const Radio = React.forwardRef<HTMLInputElement, IRadioProps>((props, ref) => {
    const {children, className, onFocus, onBlur, disabled, labelAttributes, ...inputAttributes} = props;
    const [focusVisible, setFocusVisible] = useState(false);
    const focusSource = useRef(EFocusSource.NONE);
    const classNames = classnames('cssClass[radio]', className);
    const labelClassNames = classnames(
        'cssClass[label]',
        {'cssClass[disabled]': !!disabled, 'cssClass[nonempty]': !!children},
        labelAttributes?.className
    );
    const inputRef = useRef<HTMLInputElement | null>(null);

    /** Обработчик клика. */
    const handleClick = (event: React.MouseEvent<HTMLLabelElement>) => {
        if (!disabled) {
            if (event.target === inputRef.current) {
                // Сбрасываем состояние, если событие пришло от радио-кнопки не в фокусе. (Safari)
                if (event.target !== document.activeElement) {
                    focusSource.current = EFocusSource.NONE;
                }
            } else if (focusSource.current === EFocusSource.NONE) {
                focusSource.current = EFocusSource.MOUSE;
            }
        }
        labelAttributes?.onClick?.(event);
    };

    /** Обработчик нажатия мыши. */
    const handleMouseDown = (event: React.MouseEvent<HTMLLabelElement>) => {
        if (!disabled) {
            if (focusSource.current === EFocusSource.NONE) {
                focusSource.current = EFocusSource.MOUSE;
            }
        }
        labelAttributes?.onMouseDown?.(event);
    };

    /** Обработчик получения фокуса. */
    const handleFocus = (event: React.FocusEvent<HTMLInputElement>) => {
        if (focusSource.current !== EFocusSource.MOUSE) {
            focusSource.current = EFocusSource.KEYBOARD;
            setFocusVisible(true);
        } else if (focusVisible) {
            setFocusVisible(false);
        }
        onFocus?.(event);
    };

    /** Обработчик потери фокуса. */
    const handleBlur = (event: React.FocusEvent<HTMLInputElement>) => {
        if (focusSource.current !== EFocusSource.NONE) {
            // При переключении окон/вкладок состояние не сбрасываем.
            if (event.target !== document.activeElement) {
                focusSource.current = EFocusSource.NONE;
                setFocusVisible(false);
            }
        }
        onBlur?.(event);
    };

    /** Функция для хранения ссылки. */
    const setRef = (instance: HTMLInputElement | null) => {
        inputRef.current = instance;
        if (typeof ref === 'function') {
            ref(instance);
        } else if (ref) {
            ref.current = instance;
        }
    };

    return (
        <label
            {...labelAttributes}
            className={labelClassNames}
            onClick={handleClick}
            onMouseDown={handleMouseDown}
            data-tx={process.env.npm_package_version}
        >
            <input
                type="radio"
                className={classNames}
                disabled={disabled}
                data-focus-visible={focusVisible ? '' : undefined}
                onFocus={handleFocus}
                onBlur={handleBlur}
                {...inputAttributes}
                ref={setRef}
            />
            <span className="cssClass[radioIcon]" />
            {children}
        </label>
    );
});

Radio.displayName = 'Radio';
