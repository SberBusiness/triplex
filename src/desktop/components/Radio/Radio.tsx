import React, {useState, useRef} from 'react';
import {IRadioProps} from '@sberbusiness/triplex/desktop/components/Radio/types';
import {Text} from '@sberbusiness/triplex/desktop/components/Typography/Text';
import {ELineType, ETextSize} from '@sberbusiness/triplex/desktop/components/Typography/enums';
import {classnames} from '@sberbusiness/triplex/common/utils/classnames/classnames';
import {EFocusSource} from '@sberbusiness/triplex/common/enums/EFocusSource';

/** Радио-кнопка с описанием. */
export const Radio = React.forwardRef<HTMLInputElement, IRadioProps>((props, ref) => {
    const {children, className, onFocus, onBlur, disabled, labelAttributes, ...inputAttributes} = props;
    const [focusVisible, setFocusVisible] = useState(false);
    const focusSource = useRef(EFocusSource.NONE);
    const classNames = classnames('cssClass[radio]', className);
    const labelClassNames = classnames(
        'cssClass[label]',
        {'cssClass[enabled]': !disabled, 'cssClass[nonempty]': !!children},
        labelAttributes?.className
    );
    const inputRef = useRef<HTMLInputElement | null>(null);

    /** Обработчик нажатия мыши. */
    const handleMouseDown = (event: React.MouseEvent<HTMLLabelElement>) => {
        if (!disabled) {
            if (focusSource.current === EFocusSource.NONE) {
                focusSource.current = EFocusSource.MOUSE;
            }
        }
        labelAttributes?.onMouseDown?.(event);
    };

    /** Обработчик клика. */
    const handleClick = (event: React.MouseEvent<HTMLLabelElement>) => {
        if (!disabled) {
            if (event.target === inputRef.current) {
                // Сбрасываем состояние, если событие пришло от радио-кнопки не в фокусе. (Safari)
                if (event.target !== document.activeElement) {
                    focusSource.current = EFocusSource.NONE;
                }
            } else if (focusSource.current === EFocusSource.NONE) {
                focusSource.current = EFocusSource.MOUSE;
            }
        }
        labelAttributes?.onClick?.(event);
    };

    /** Обработчик получения фокуса. */
    const handleFocus = (event: React.FocusEvent<HTMLInputElement>) => {
        if (focusSource.current !== EFocusSource.MOUSE) {
            focusSource.current = EFocusSource.KEYBOARD;
            setFocusVisible(true);
        } else if (focusVisible) {
            setFocusVisible(false);
        }
        onFocus?.(event);
    };

    /** Обработчик потери фокуса. */
    const handleBlur = (event: React.FocusEvent<HTMLInputElement>) => {
        if (focusSource.current !== EFocusSource.NONE) {
            // При переключении окон/вкладок состояние не сбрасываем.
            if (event.target !== document.activeElement) {
                focusSource.current = EFocusSource.NONE;
                setFocusVisible(false);
            }
        }
        onBlur?.(event);
    };

    /** Функция для хранения ссылки. */
    const setRef = (instance: HTMLInputElement | null) => {
        inputRef.current = instance;
        if (typeof ref === 'function') {
            ref(instance);
        } else if (ref) {
            ref.current = instance;
        }
    };

    return (
        <Text
            tag="label"
            size={ETextSize.B1}
            line={ELineType.EXTRA}
            {...labelAttributes}
            className={labelClassNames}
            onMouseDown={handleMouseDown}
            onClick={handleClick}
        >
            <input
                type="radio"
                className={classNames}
                onFocus={handleFocus}
                onBlur={handleBlur}
                disabled={disabled}
                data-focus-visible={focusVisible ? '' : undefined}
                ref={setRef}
                {...inputAttributes}
            />
            <span className="cssClass[radioIcon]" />
            {children}
        </Text>
    );
});

Radio.displayName = 'Radio';
