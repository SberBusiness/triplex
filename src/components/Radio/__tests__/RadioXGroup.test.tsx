import React from 'react';
import renderer from 'react-test-renderer';
import {allure} from '@jest/unit/allure-report';
import {RadioXGroup} from '@sberbusiness/triplex/components/Radio/RadioXGroup';
import {Radio} from '@sberbusiness/triplex/components/Radio/Radio';

describe('RadioXGroup', () => {
    beforeEach(() => {
        allure.feature('RadioXGroup');
    });

    it('renders correctly', () => {
        const tree = renderer
            .create(
                <RadioXGroup>
                    <Radio>Sample Text</Radio>
                    <Radio>Sample Text</Radio>
                    <Radio>Sample Text</Radio>
                </RadioXGroup>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
