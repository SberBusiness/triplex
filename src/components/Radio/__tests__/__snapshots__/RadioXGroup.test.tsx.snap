// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioXGroup renders correctly 1`] = `
<div
  className="cssClass[radioXGroup] cssClass[indent-12]"
  role="radiogroup"
>
  <label
    className="cssClass[label] cssClass[nonempty]"
    onClick={[Function]}
    onMouseDown={[Function]}
  >
    <input
      className="cssClass[radio]"
      onBlur={[Function]}
      onFocus={[Function]}
      type="radio"
    />
    <span
      className="cssClass[radioIcon]"
    />
    Sample Text
  </label>
  <label
    className="cssClass[label] cssClass[nonempty]"
    onClick={[Function]}
    onMouseDown={[Function]}
  >
    <input
      className="cssClass[radio]"
      onBlur={[Function]}
      onFocus={[Function]}
      type="radio"
    />
    <span
      className="cssClass[radioIcon]"
    />
    Sample Text
  </label>
  <label
    className="cssClass[label] cssClass[nonempty]"
    onClick={[Function]}
    onMouseDown={[Function]}
  >
    <input
      className="cssClass[radio]"
      onBlur={[Function]}
      onFocus={[Function]}
      type="radio"
    />
    <span
      className="cssClass[radioIcon]"
    />
    Sample Text
  </label>
</div>
`;
